src/vs/workbench/contrib/notebook/browser/controller/sectionActions.ts

```diff
@@ -0,0 +1,210 @@
+			},
+			shortTitle: localize('foldSection', "Fold Section"),
+			menu: [
+				{
+					id: MenuId.NotebookOutlineActionMenu,
+					group: 'notebookFolding',
+					order: 2,
+					when: ContextKeyExpr.and(
+						NotebookOutlineContext.CellKind.isEqualTo(CellKind.Markup),
+						NotebookOutlineContext.OutlineElementTarget.isEqualTo(OutlineTarget.OutlinePane),
+						NotebookOutlineContext.CellHasChildren,
+						NotebookOutlineContext.CellHasHeader,
+						NotebookOutlineContext.CellFoldingState.isEqualTo(CellFoldingState.Expanded)
+					)
+				}
+			]
+		});
+	}
+
+	override async run(_accessor: ServicesAccessor, context: NotebookSectionArgs): Promise<void> {
+		if (!checkSectionContext(context)) {
+			return;
+		}
+
+		this.toggleFoldRange(context.outlineEntry, context.notebookEditor!);
+	}
+
+	private toggleFoldRange(entry: OutlineEntry, notebookEditor: INotebookEditor) {
+		const foldingController = notebookEditor.getContribution<FoldingController>(FoldingController.id);
+		const index = entry.index;
+		const headerLevel = entry.level;
+		const newFoldingState = CellFoldingState.Collapsed;
+
+		foldingController.setFoldingStateDown(index, newFoldingState, headerLevel);
+	}
+}
+
+export class NotebookExpandSection extends Action2 {
+	constructor() {
+		super({
+			id: 'notebook.section.expandSection',
+			title: {
+				...localize2('expandSection', "Expand Section"),
+				mnemonicTitle: localize({ key: 'miexpandSection', comment: ['&& denotes a mnemonic'] }, "&&Expand Section"),
+			},
+			shortTitle: localize('expandSection', "Expand Section"),
+			menu: [
+				{
+					id: MenuId.NotebookOutlineActionMenu,
+					group: 'notebookFolding',
+					order: 2,
+					when: ContextKeyExpr.and(
+						NotebookOutlineContext.CellKind.isEqualTo(CellKind.Markup),
+						NotebookOutlineContext.OutlineElementTarget.isEqualTo(OutlineTarget.OutlinePane),
+						NotebookOutlineContext.CellHasChildren,
+						NotebookOutlineContext.CellHasHeader,
+						NotebookOutlineContext.CellFoldingState.isEqualTo(CellFoldingState.Collapsed)
+					)
+				}
+			]
+		});
+	}
+
+	override async run(_accessor: ServicesAccessor, context: NotebookSectionArgs): Promise<void> {
+		if (!checkSectionContext(context)) {
+			return;
+		}
+
+		this.toggleFoldRange(context.outlineEntry, context.notebookEditor!);
+	}
+
+	private toggleFoldRange(entry: OutlineEntry, notebookEditor: INotebookEditor) {
+		const foldingController = notebookEditor.getContribution<FoldingController>(FoldingController.id);
+		const index = entry.index;
+		const headerLevel = entry.level;
+		const newFoldingState = CellFoldingState.Expanded;
+
+		foldingController.setFoldingStateDown(index, newFoldingState, headerLevel);
+	}
+}
+
+/**
+ * Take in context args and check if they are valid
+ *
+ * @param context - Notebook Section Context containing a notebook editor and outline entry
+ * @returns true if context is valid, false otherwise
+ */
+function checkSectionContext(context: NotebookSectionArgs): boolean {
```

Comment:
To avoid using `!` everywhere, what we can do is

```
export type ValidNotebookSectionArgs = {
	notebookEditor: INotebookEditor;
	outlineEntry: OutlineEntry;
};

function checkSectionContext(context: NotebookSectionArgs | undefined): context is ValidNotebookSectionArgs {
	return !!(context && context.notebookEditor && context.outlineEntry);
}
```